### ISSUE TITLE
PyMC 6 / ArviZ 1.0 / jax-0.10 migration + SBI backend replacement (sbijax → pyabc + BayesFlow + sbi)

### ISSUE THREAD

**Issue by jhuggins** [TARPS-group/prob-pipe#238]
## Summary

Two coupled workstreams, both unblocked by the same action — **dropping `sbijax`**:

- **(A) Ecosystem upgrade** to PyMC 6 / PyTensor 3 / ArviZ 1.0 / jax-0.10. **Empirically verified:** a clean target env runs the full test suite green (**2807 passed, 8 skipped, 0 failed**) after a **single one-line source fix**. The upgrade's code surface is ~1 line plus packaging.
- **(B) SBI backend replacement** — retire `sbijax` (single maintainer, `jax==0.8.1`-pinned, dm-haiku / tfp-nightly tail) and replace its functionality with better-maintained backends.

The two are **orthogonal** and parallelizable once sbijax is gone. `sbijax` is the *sole* blocker to a single co-installable `jax 0.10 / pymc 6 / arviz 1.x` environment (it hard-pins `jax==0.8.1` and `arviz==1.0.0`, which cannot co-resolve with pymc6's `arviz>=1.1`).

## Verified target environment

`pymc 6.0.1 · pytensor 3.0.4 · numba 0.65.1 · arviz 1.1.0 · jax/jaxlib 0.10.1 · blackjax 1.5 · tfp-nightly (dev20260605) · nutpie 0.16.10`. The only required source change is `pm.sample_prior_predictive(samples=)` → `draws=` (`probpipe/modeling/_pymc.py`).

## SBI backend ownership (decided)

| Capability | Backend | Install |
|---|---|---|
| SMC-ABC | **pyabc** | **core** — guarantees basic SBI on a bare `pip install probpipe` |
| NPE / FMPE / CMPE (amortized) | **BayesFlow** (keras-on-jax) | `[bayesflow]` extra (jax-native, no torch) |
| NLE / NRE (+ MCMC) | **sbi** (PyTorch) | `[sbi-torch]` extra (heavy) |

Rationale: pyabc is **core** so basic SBI ships by default (and its local-multicore parallelism comes with it); BayesFlow is an **extra** because core-ifying it would cap the *whole library* at Python `<3.14` and force `numpy>=2.2.6` on every user; PyTorch is confined to NLE/NRE, the only methods that need it. NLE/NRE use sbi's **torch-native** sampler — `slice_np_vectorized` by default, gradient-based NUTS opt-in behind a **hard correctness guard** (a gradient sampler + a non-torch-differentiable prior otherwise yields a silently-wrong, likelihood-only gradient).

## Roadmap (tracking)

- [x] **0 — Drop sbijax.** Remove the extra + `_sbijax.py` + `test_sbijax.py`; delete the CI `--no-deps sbijax` hacks + stale comment; guard the SBI notebook cells. Unblocks the jax-0.10/pymc6 env; zero replacement code.
- [x] **1 — jax/jaxlib + blackjax bump** (own PR; highest RNG/numerics blast radius — isolate for bisectability).
- [x] **2 — arviz `<1.0`→`<2.0` ceiling lift**; keep the dual-version straddle through Step 7; bind `arviz_*` by name (incl. `_cmdstan_method.py` + a `[stan]` smoke test); add `[diagnostics]` extra. **Hard cutover** — CI tests the arviz 1.x line only.
- [x] **3 — `[pymc]`→`pymc>=6`** + CI leg (3.12 & 3.13); apply the `draws=` fix; add matplotlib to the env; reclaim pymc6 multi-core (relax `cores=1`, verify fork-safety).
- [x] **4 — pyabc SMC-ABC backend** (core dep; register `pyabc_smcabc`; eliminates the sbijax 1D-param bug + chol-factor monkeypatch). Prereq: extend `make_posterior` to forward `weights=`.
- [x] **5 — BayesFlow backend** (`[bayesflow]`; amortized NPE/FMPE/CMPE as direct samplers + jax summary nets; `KERAS_BACKEND=jax`; CI on 3.12/3.13 only).
- [ ] **6 — sbi torch backend** (`[sbi-torch]`; NLE/NRE, slice default, guarded gradient path, CPU torch wheel in CI). **Run a resolver dry-run + recipe smoke test first.**
- [x] **6a — (spike) jax-native NLE via BayesFlow + BlackJAX** — if its keras-jax `log_prob` proves `jax.grad`-transparent, torch drops out of the default surface entirely. Gates Step 6's scope.
- [x] **7 — Diagnostics (PR #193)** — redesign to pure ops returning `Record`s behind one private arviz bridge; then port to arviz 1.0 + a frozen-fixture regression suite; adopt the new 0.89/eti defaults (update example numbers).
- [x] **8 — Straddle deletion** + raise core to `arviz>=1.1,<2.0` (intentionally drops arviz-0.x support) + bridge simplification.
- [x] **9 — (optional) `probpipe-core` two-distribution split** for a lean install — see #237. ✅ **Mechanism landed in #290** (`probpipe-core` lean + code-less `probpipe` metapackage). Note: pyabc is not in the metapackage yet — it joins once Step 4 (pyabc backend) lands.

## Key risks / guards

- **Silent wrong posterior**: gradient sampler + non-torch-differentiable prior → likelihood-only gradient, no error. **Hard construction-time guard** required (never silently build the numpy-prior `CustomPriorWrapper` path).
- **numpy window**: core `>=2.0,<2.5`; `[bayesflow]` tightens to `>=2.2.6,<2.5` — pin/track in CI.
- **torch footprint**: the default Linux wheel pulls ~3 GB CUDA → CI must use `--extra-index-url .../whl/cpu`; CPU-torch is the SBI default.
- **Unverified-by-install**: the torch+jax co-install and the sbi NLE→`MCMCPosterior` recipe are validated from constraint metadata / source, not yet a live install — smoke-test before Step 6.

## Resolved decisions

arviz → adopt new 0.89/eti defaults (not called in source; update examples) · pyabc → **core** · BayesFlow → **extra**, not core (Python `<3.14` cap) · sbi NLE/NRE default sampler `slice_np_vectorized`, NUTS opt-in · jax-native-NLE spike → **gated** (Step 6a before Step 6) · `cores=1` → reclaim multi-core · arviz → **hard cutover** (no dual-version test window).

## Packaging (pyproject)

```
core:        jax, jaxlib, tfp-nightly, blackjax, arviz, pyabc
[bayesflow]  bayesflow>=2.0,<3         # amortized neural SBI, jax-native (numpy>=2.2.6, Python <3.14)
[sbi-torch]  sbi[pyro]>=0.26,<0.27     # NLE/NRE, pulls torch + pyro
[sbi]        = [bayesflow, sbi-torch]  # full neural SBI on top of core pyabc
[pymc] [nutpie] [diagnostics] [stan]
```

Sequencing is orthogonal to the field-order stack (`#234`/`#236`) except Step 8's bridge-simplification half. Steps 4 and 5 are mutually independent; Step 6 follows the 6a spike.

**Comment by jhuggins** [comment]
**Status: Step 4 (pyabc) is unblocked and starting now** — assigning to @XinyueLi1012.

**Not gated on the pymc-upgrade path.** pyabc is pure-Python and JAX-agnostic — it co-installs with current `main` (it never needed jax 0.10 / PyMC 6 / ArviZ 1.0), so it runs in **parallel** with Steps 0–3 rather than after them.

**Only prerequisite:** extend `make_posterior(...)` to forward a `weights=` kwarg to `ApproximateDistribution` (a small additive core change) so SMC-ABC's weighted particles flow through. Independent of the migration.

**Soft note (not a blocker):** when #235's Phase-4 renames land (`*Array`→`*Batch`, etc.), pyabc's code absorbs them in the same mechanical sweep — coordinate, but don't wait.

Scope reminder for this step:
- register `pyabc_smcabc` (`supported_types=(SimpleGenerativeModel,)`, priority 6) in the existing `inference_method_registry` — no new registry;
- in-memory `"sqlite://"` history; weighted particles via the new `weights=` path;
- new `tests/inference/test_pyabc.py` (SMC-ABC auto-dispatch via `condition_on`; posterior recovery near truth); re-baseline to calibration-style assertions;
- eliminates the sbijax `_chol_factor` 1D monkeypatch and the dim≥2 caveat.

The pyabc-into-core-`dependencies` packaging line rides with the #238 packaging work; the backend implementation + tests proceed now regardless.

**Comment by jhuggins** [comment]
Status check: every committed step here has landed — #240–#243 (the PyMC 6 / ArviZ 1.0 / jax upgrades), #293 (pyabc), #246 (BayesFlow), #259 (jax-native NLE/NRE), #193 (diagnostics), and the #290 core split. The only unchecked scope is the optional torch-backed `sbi` integration; worth retitling/trimming this issue to that remainder if we decide to pursue it.
